### PR TITLE
Added ir.Optimize

### PR DESCRIPTION
**Pull request by cseed**
Mostly code reorg.  Also:

moved rewriters to ir objects
call Optimize before intepreting
removed Filter{Rows, Cols} rules (non-IR), those should get folded back into the MT methods like other AST-based rules
re-enabled Fitler{Rows, Cols}IR fusion rules since logical and/or is back

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/Interpret.scala [hunk 1]
```diff
@@ -14,18 +14,22 @@ object Interpret {
...
     agg: Option[Agg]): T = {
 
-    log.info("interpret:\n" + Pretty(ir))
 
     val (typeEnv, valueEnv) = env.m.foldLeft((Env.empty[Type], Env.empty[Any])) {
       case ((e1, e2), (k, (value, t))) => (e1.bind(k, t), e2.bind(k, value))
     }
 
+    var ir = ir0
+    ir = Optimize(ir)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/Interpret.scala]: should this be parameterized? If we want to use the interpreter to test our compiler and optimizer, then we need a way to take the stupid path

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]: Yeah, agreed, will change.

Comment (tpoterba) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]: 🤔 🕐

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]: Don't ask :-/

**Pull request merged**